### PR TITLE
T5 & mT5

### LINKED ISSUE #6285: 🌟 T5 V1.1
# 🌟 New model addition

## Model description

T5 version t5.1.1.* is very similar to the original T5 model, with the following differences:
- GEGLU activation in feed-forward hidden layer, rather than ReLU - see https://arxiv.org/abs/2002.05202 .
- Dropout was turned off in pre-training (quality win). Dropout should be re-enabled during fine-tuning.
- Pre-trained on C4 only without mixing in the downstream tasks.
- no parameter sharing between embedding and classifier layer
- "xl" and "xxl" replace "3B" and "11B". The model shapes are a bit different - larger d_model and smaller num_heads and d_ff.

The key reason why these models are interesting is that - unlike the originally released models - they were trained **only** on unlabeled data and not on any labeled data, making them applicable for few-shot learning experiments. As they are very similar to the original T5 models, I assume they are relatively easy to implement.

## Open source status

* [x] the model implementation is available: (give details) - see https://github.com/google-research/text-to-text-transfer-transformer/
* [x] the model weights are available: (give details) - see https://github.com/google-research/text-to-text-transfer-transformer/blob/master/released_checkpoints.md
* [x] who are the authors: (mention them, if possible by @gh-username) - Colin Raffel ( @craffel ), Noam Shazeer ( @nshazeer ), Adam Roberts ( @adarob ), Katherine Lee, Sharan Narang, Michael Matena ( @mmatena ), Yanqi Zhou, Wei Li, Peter J. Liu

(Also tagging @patrickvonplaten as he is mentioned in the **who to tag** guide for T5)

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds T5v1.1 and MT5.

I was really unsure whether I should make a new model file `modeling_5_v1_1.py` or not. I finally decided (also after discussions in https://github.com/huggingface/transformers/issues/6285) that in this case, it is better to add the few new T5 features to the existing `modeling_t5.py` file.

The context is the following:
The T5 team released weights for a **T5v1.1** model: https://github.com/google-research/text-to-text-transfer-transformer/blob/master/released_checkpoints.md. These model architecture of T5v1.1 is equivalent to T5 besides two changes:
1) input and output word embedding of decoder is not shared anymore
2) different feed forward layer is used
In addition, the **mT5** model checkpoints were also released and are fully based on T5v1_1: https://github.com/google-research/multilingual-t5#released-model-checkpoints .

Now the philosophy of the library is to create a new model class if the architecture slightly differs from a previously integrated model, but I'd argue that in this case it is better to add a new config param called "feed_forward_proj" which defines whether a "relu" (T5) or a "gated-gelu" (T5v1.1) should be used. The arguments for not creating a new model class are the following:

1) Both T5 and T5v1.1 are "officially" the same model and both belong to the same code base: https://github.com/google-research/text-to-text-transfer-transformer
2) T5v1.1 has no official paper and it's quite difficult to find a good name as discussed in https://github.com/huggingface/transformers/issues/6285 => `T5v11` is a very cryptic name IMO and `T5v2` is not ideal either.
3) One could argue that it makes sense to create a mixture of T5 and T5v1v1 by sharing input and output word embedding of the decoder but using the "gated-gelu" feed-forward layer. This design would allow a user to define a model whereas creating a new T5v11 model  file would make this impossible
4) `"feed-forward-proj"` is less of a model-specific architecture configuration than `"do_blenderbot_90_layernorm"` IMO

The obvious disadvantage is that I am adding new architecture code to an existing model to make a new model (or new model version) work, which is kinda what we did not want to do. 

It's a tough decision here I think, so I'd be very happy for your feedback on this @LysandreJik @sgugger @thomwolf . I'd also understand if you guys think we should make a new model file here.

I already uploaded `google/t5-v1_1-small` and `google/mt5-small`. If you guys are ok with this PR, I'll add the TF code, add the new model to README.md and we're good for merge.

### CODE REVIEW

**Comment by thomwolf** [comment]
I'm in favor of this solution and happy you selected it :)

Our philosophy of simple to use and modify standalone code is made not to be a hard requirement but a guide for user-experience and in this case I also think the best is to adapt T5 as you do.

**Comment by nzinov** [comment]
Big thanks for integrating mT5! I checked out your PR and there seems to be a problem with extra_ids. I guess, sentencepiece vocab already has <extra_id> tokens but T5Tokenizer goes on to add it's own.  E.g., real id of <external_id_0> is 250099 but T5Tokenizer translates it into 250199

**Comment by patrickvonplaten** [comment]
> Big thanks for integrating mT5! I checked out your PR and there seems to be a problem with extra_ids. I guess, sentencepiece vocab already has <extra_id> tokens but T5Tokenizer goes on to add it's own. E.g., real id of <external_id_0> is 250099 but T5Tokenizer translates it into 250199

You're 100% correct - thanks a lot for letting me know! I corrected the behavior - should be good now: https://huggingface.co/google/mt5-small/tree/main

File: src/transformers/configuration_mt5.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+from .utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+
+class MT5Config(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a :class:`~transformers.MT5Model` or a
+    :class:`~transformers.TFMT5Model`. It is used to instantiate a mT5 model according to the specified arguments,
```
Comment (sgugger) [inline src/transformers/configuration_mt5.py:27]: There is no `TFMT5Model` (yet?). We can leave as is if you think you'll add it very soon, otherwise this part should be removed.

File: src/transformers/configuration_mt5.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    defining the model architecture. Instantiating a configuration with the defaults will yield a similar configuration
+    to that of the mT5 `google/mt5-small <https://huggingface.co/google/mt5-small>`__ architecture.
+
+    Configuration objects inherit from :class:`~transformers.PretrainedConfig` and can be used to control the model
+    outputs. Read the documentation from :class:`~transformers.PretrainedConfig` for more information.
+
+    Arguments:
+        vocab_size (:obj:`int`, `optional`, defaults to 32128):
+            Vocabulary size of the T5 model. Defines the number of different tokens that can be represented by the
+            :obj:`inputs_ids` passed when calling :class:`~transformers.T5Model` or :class:`~transformers.TFT5Model`.
```
Comment (sgugger) [inline src/transformers/configuration_mt5.py:37]: Same comment here.

File: src/transformers/configuration_mt5.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            The number of buckets to use for each attention layer.
+        dropout_rate (:obj:`float`, `optional`, defaults to 0.1):
+            The ratio for all dropout layers.
+        layer_norm_eps (:obj:`float`, `optional`, defaults to 1e-6):
+            The epsilon used by the layer normalization layers.
+        initializer_factor (:obj:`float`, `optional`, defaults to 1):
+            A factor for initializing all weight matrices (should be kept to 1, used internally for initialization
+            testing).
+        feed_forward_proj (:obj:`string`, `optional`, defaults to :obj:`"gated-gelu"`):
+            Type of feed forward layer to be used. Should be one of :obj:`"relu"` or :obj:`"gated-gelu"`. T5v1.1 and
```
Comment (sgugger) [inline src/transformers/configuration_mt5.py:62]: Since we are in the doc of mT5, I would remove the last sentence entirely. (Documentation is not a Choose your own adventure book.)

File: src/transformers/modeling_mt5.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+logger = logging.get_logger(__name__)
+
+_CONFIG_FOR_DOC = "T5Config"
+_TOKENIZER_FOR_DOC = "T5Tokenizer"
+
+####################################################
+# This dict contains shortcut names and associated url
+# for the pretrained weights provided with the models
+####################################################
```
Comment (sgugger) [inline src/transformers/modeling_mt5.py:30]: We usually remove those comments.

File: src/transformers/configuration_t5.py [hunk 5]
```diff
@@ -69,6 +66,9 @@ class T5Config(PretrainedConfig):
         initializer_factor (:obj:`float`, `optional`, defaults to 1):
             A factor for initializing all weight matrices (should be kept to 1, used internally for initialization
             testing).
+        feed_forward_proj (:obj:`string`, `optional`, defaults to :obj:`"relu"`):
+            Type of feed forward layer to be used. Should be one of :obj:`"relu"` or :obj:`"gated-gelu"`. T5v1.1 and
+            mT5 use the :obj:`"gated-gelu"` feed forward projection. Original T5 uses :obj:`"relu"`.
```
Comment (sgugger) [inline src/transformers/configuration_t5.py:71]: Same here, remove mention of mT5 and only speak about T5 and T5v1.1.

**Review by sgugger — APPROVED** [review]
Looks very clean! I agree with your design choice, for all the reasons you mentioned.

File: src/transformers/configuration_mt5.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/configuration_mt5.py:27]: I'll add `TFMT5Model` now :-)

File: src/transformers/modeling_mt5.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+T5_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "google/t5-small",
+    "google/t5-base",
+    "google/t5-large",
+    "google/t5-3b",
+    "google/t5-11b",
+    # See all T5 models at https://huggingface.co/models?filter=t5
+]
```
Comment (LysandreJik) [inline src/transformers/modeling_mt5.py:38]: Is this needed here? If it is, it probably shouldn't be those model identifiers but `mT5-{Small, Base, Large, XL, XXL}` instead

File: src/transformers/configuration_mt5.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
+# coding=utf-8
+# Copyright 2010, The T5 Authors and HuggingFace Inc.
```
Comment (LysandreJik) [inline src/transformers/configuration_mt5.py:2]: Wow that's an old copyright

**Review by LysandreJik — APPROVED** [review]
Thank you for the extremely detailed description of why you made the choice you did. Definitely agree with you that in this case it can be added to the existing modeling T5 files.

LGTM, clean implementation.

**Comment by patrickvonplaten** [comment]
Ran the slow tests & added multiple new slow tests => good to merge.

File: src/transformers/__init__.py [hunk 8]
```diff
@@ -498,6 +498,7 @@
         MobileBertPreTrainedModel,
         load_tf_weights_in_mobilebert,
     )
+    from .models.mt5 import MT5Config, MT5ForConditionalGeneration, MT5Model
```
Comment (sgugger) [inline src/transformers/__init__.py:501]: The config should be always imported, not just under `is_torch_available`.
